### PR TITLE
Allow replace a never_connected agent

### PR DESCRIPTION
**Pull request by MiguelazoDS**
|Related issue|
|---|
|#10185|

## Description

This PR adds the capability to replace a never_connected agent since it could be in that state indefinitely and will never be replaced.

## Logs/Alerts example
- Agent is not replaced since it was not disconnected long enough.
  

  
  
  
- Agent is not replaced since it is connected.
  
- Agent is replaced.
  
  
  
  
  
- Never connected agent is replaced as well.
  

- Compilation without warnings in every supported platform
  - [X] Linux
- [X] Source installation
- [X] Review logs syntax and correct language

### CODE REVIEW

File: src/addagent/validate.c [hunk 1]
```diff
@@ -550,9 +550,15 @@ double get_time_since_agent_disconnection(const char *id) {
         status = json_field->valuestring;
     }
 
-    json_field = cJSON_GetObjectItem(json_agent_info->child, "disconnected_time");
-    if (json_field && cJSON_IsNumber(json_field) && status && !strcmp(status, AGENT_CS_DISCONNECTED)) {
-        disconnected_time = json_field->valueint;
+    if (status) {
+        if (!strcmp(status, AGENT_CS_NEVER_CONNECTED)) {
+            return OS_INVALID;
+        } else if (!strcmp(status, AGENT_CS_DISCONNECTED)) {
```
Comment (palaciosjeremias) [inline src/addagent/validate.c:556]: Code style:
Use the **else if** statemente in a new line.

File: src/addagent/validate.c [hunk 2]
```diff
@@ -550,9 +550,15 @@ double get_time_since_agent_disconnection(const char *id) {
         status = json_field->valuestring;
     }
 
-    json_field = cJSON_GetObjectItem(json_agent_info->child, "disconnected_time");
-    if (json_field && cJSON_IsNumber(json_field) && status && !strcmp(status, AGENT_CS_DISCONNECTED)) {
-        disconnected_time = json_field->valueint;
+    if (status) {
+        if (!strcmp(status, AGENT_CS_NEVER_CONNECTED)) {
+            return OS_INVALID;
```
Comment (palaciosjeremias) [inline src/addagent/validate.c:555]: Memory leak: This return will leave **json_agent_info** allocated generating a memory leak.

**Review by palaciosjeremias — CHANGES_REQUESTED** [review]
I think we must implement a different approach here:
Returning **OS_INVALID** works because **w_auth_replace_agent()** isn´t controlling if the result is valid. But it should, otherwise, if WazuhDB is down, we will allow replacing agents.
In example, we can implement a method **check_agent_disconnection_time()**, where we will approve if the agent is disconnected for enough time or because it's **never_connected**

File: src/addagent/validate.c (hunk 1 above)
Comment (MiguelazoDS) [reply src/addagent/validate.c:556]: According to this convention, we are using the else clause after the "}" in the same line. 

https://github.com/wazuh/wazuh/wiki/Coding-style-guide-%28C%29#style-conventions

Comment (palaciosjeremias) [reply src/addagent/validate.c:556]: You are right

**Comment by MiguelazoDS** [comment]
`valgrind --show-leak-kinds=all --leak-check=full --track-origins=yes -s ./test_auth_validate`
<details>
  <summary>Valgrind report test_auth_validate</summary>

==289882== Memcheck, a memory error detector
==289882== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==289882== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==289882== Command: ./test_auth_validate
==289882== 
[==========] Running 10 test(s).
[ RUN      ] test_w_auth_validate_groups
[       OK ] test_w_auth_validate_groups
[ RUN      ] test_w_auth_validate_data
[       OK ] test_w_auth_validate_data
[ RUN      ] test_w_auth_validate_data_force_insert
[       OK ] test_w_auth_validate_data_force_insert
[ RUN      ] test_w_auth_replace_agent_force_disabled
[       OK ] test_w_auth_replace_agent_force_disabled
[ RUN      ] test_w_auth_replace_agent_not_comply_antiquity
[       OK ] test_w_auth_replace_agent_not_comply_antiquity
[ RUN      ] test_w_auth_replace_agent_not_disconnected_long_enough
[       OK ] test_w_auth_replace_agent_not_disconnected_long_enough
[ RUN      ] test_w_auth_replace_agent_not_disconnected
[       OK ] test_w_auth_replace_agent_not_disconnected
[ RUN      ] test_w_auth_replace_agent_not_old_enough
[       OK ] test_w_auth_replace_agent_not_old_enough
[ RUN      ] test_w_auth_replace_agent_existent_key_hash
[       OK ] test_w_auth_replace_agent_existent_key_hash
[ RUN      ] test_w_auth_replace_agent_success
[       OK ] test_w_auth_replace_agent_success
[==========] 10 test(s) run.
[  PASSED  ] 10 test(s).
==289882== 
==289882== HEAP SUMMARY:
==289882==     in use at exit: 0 bytes in 0 blocks
==289882==   total heap usage: 1,838 allocs, 1,838 frees, 210,241 bytes allocated
==289882== 
==289882== All heap blocks were freed -- no leaks are possible
==289882== 
==289882== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
</details>

File: src/addagent/validate.c (hunk 2 above)
Comment (MiguelazoDS) [reply src/addagent/validate.c]: You're right

File: src/os_auth/auth.c [hunk 3]
```diff
@@ -202,24 +209,55 @@ w_err_t w_auth_replace_agent(keyentry *key,
...
+        if (!cJSON_IsString(j_connection_status)) {
+            mdebug1("Failed to get connection status for agent '%s'.", key->id);
+            cJSON_Delete(j_agent_info);
             return OS_INVALID;
-        } else if (disconnected_time > 0 && disconnected_time < force_options->disconnected_time) {
-            minfo("Agent '%s' has not been disconnected long enough to be replaced.", key->id);
+        }
+        if (!cJSON_IsNumber(j_disconnected_time)) {
+            mdebug1("Failed to get disconnected time for agent '%s'.", key->id);
+            cJSON_Delete(j_agent_info);
```
Comment (palaciosjeremias) [inline src/os_auth/auth.c:230]: We should avoid all this duplicated cJSON_Delete(j_agent_info);
I suggest you declaring a boolean **replace_agent** initialized to true. And you set it to false on each return **OS_INVALID**, at the end of the function you check this value to eliminate or not the old agent.

File: src/os_auth/auth.c [hunk 4]
```diff
@@ -202,24 +209,55 @@ w_err_t w_auth_replace_agent(keyentry *key,
...
+    if (force_options->after_registration_time > 0) {
+        if (!cJSON_IsNumber(j_date_add)) {
+            mdebug1("Failed to get registration time for agent '%s'.", key->id);
+            cJSON_Delete(j_agent_info);
+            return OS_INVALID;
+        }
+
+        time_t agent_registration_time = difftime(time(NULL), j_date_add->valueint);
+
+        if (agent_registration_time > 0 && agent_registration_time < force_options->after_registration_time) {
```
Comment (palaciosjeremias) [inline src/os_auth/auth.c:258]: I think it isn´t required to check **agent_registration_time > 0**.
We can rename **agent_registration_time** to **agent_time_since_desconnection**

**Review by palaciosjeremias — CHANGES_REQUESTED** [review]

**Review by palaciosjeremias — APPROVED** [review]
LGTM.
Good job!
For the review, the following aspects was considered:
Right behavior of the code.
Memory usage / Memory leaks
Readability and organization of the code.

**Comment by MiguelazoDS** [comment]
image
image
image
image

File: src/os_auth/auth.c [hunk 5]
```diff
@@ -202,24 +209,55 @@ w_err_t w_auth_replace_agent(keyentry *key,
...
+    j_connection_status = cJSON_GetObjectItem(j_agent_info->child, "connection_status");
+    j_disconnected_time = cJSON_GetObjectItem(j_agent_info->child, "disconnected_time");
+
     /* Check if the agent has been disconnected longer than the value specified in the configuration option*/
     if (force_options->disconnected_time_enabled) {
-        time_t disconnected_time = 0;
-        disconnected_time = (long)get_time_since_agent_disconnection(key->id);
-        if (!disconnected_time) {
-            minfo("Agent '%s' can't be replaced since it is not disconnected.", key->id);
+        if (!cJSON_IsString(j_connection_status)) {
```
[the remark this answers is not in the document]
Comment (MiguelazoDS) [reply src/os_auth/auth.c]: That's a better practice. Fixed, thanks

File: src/os_auth/auth.c (hunk 3 above)
Comment (MiguelazoDS) [reply src/os_auth/auth.c]: Changed

File: src/os_auth/auth.c (hunk 4 above)
Comment (MiguelazoDS) [reply src/os_auth/auth.c]: Changed

**Review by DProvinciani — APPROVED** [review]
Looks good!

**Pull request merged**